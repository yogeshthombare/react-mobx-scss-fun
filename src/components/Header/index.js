import React, { Component } from 'react'
import { inject, observer } from 'mobx-react'
import { extendObservable } from 'mobx'
// import AutoCompleteInput from '../../common/AutoCompleteInput'
import { NavLink, Redirect } from 'react-router-dom'
import { FaSearch } from 'react-icons/fa'
import styles from './styles.module.css'
// import LocationMenu from '../../common/LocationMenu'
// import LegalTerms from '../LegalTerms'


const Header = inject('app')(observer(class Header extends Component {
  constructor(props) {
    super(props)

    const {app: {auth, taskStore, teamStore, utilityStore}} = this.props

    extendObservable(this, {
      auth: auth,
      taskStore: taskStore,
      teamStore: teamStore,
      utilityStore: utilityStore,
      search: {
        name: '',
        valid: true
      },
      redirect: false,
      showSliderMenu: false
    })
  }

  handleLogout = () => {
    this.auth.submitLogout()
    this.redirect = '/'
  }

  onUpdateInput = (name, value) => {
    this[name].name = value
  }

  onValidate = (name, valid) => {
    this[name].valid = valid
  }

  sliderSelector = () => {
    const { user: { first_name, last_name } } = this.auth
    const { activeTeam } = this.teamStore
    const userName = `${first_name} ${last_name}`
    const initials = `${first_name.charAt(0)} ${last_name.charAt(0)}`
    return (
      <div className={styles.headerProfile}>
        <div className={styles.headerProfileImage}>
          <h2>{initials}</h2>
        </div>
        <div className={styles.headerProfileInfo}>
          <h5>{userName}</h5>
          <p>{activeTeam.name}</p>
        </div>
      </div>
    );
  }

  handleSelection = item => {
    if (item) {
      if (item.team_user_uuid) {
        this.teamStore.setActiveTeam(item)
        this.teamStore.updateTeam(item)
        this.redirect = '/dashboard'
      }

      if (item.action) {
        if (item.action === 'logout') {
          this.handleLogout()
        } else {
          item.action()
        }
      }
    }
  }

  render() {
    const version = this.auth.user.release ? `Version ${this.auth.user.release.version}` : ''
    const {teams, activeTeam} = this.teamStore

    const headerProfileMenuItems = [
      {
        name: <a href="https://www.curoglobal.com">Account</a>,
      }, {
        name: <a href="https://curosoftware.zendesk.com/hc/en-us/requests/new">Help</a>
      }, {
        name: 'Release Notes',
        action: this.utilityStore.toggleReleaseNotes,
        autoClose: true,
      }, {
        name: 'Privacy Policy',
        action: this.utilityStore.togglePrivacyPolicy,
        autoClose: true,
      }, {
        name: 'Terms of Service',
        action: this.utilityStore.toggleTermsOfService,
        autoClose: true,
      }, {
        name: <NavLink to={'/dashboard'} className={styles.sliderMenuLogout}>Logout</NavLink>,
        action: 'logout'
      }, {
        name: <div className={styles.sliderMenuVersion}>{version}</div>,
      }
    ]

    if (teams && teams.length > 1) {
      const switchCompany = {
        name: 'Switch company',
        items: teams.map(team => ({...team, active: team.name === activeTeam.name}))
      }
      headerProfileMenuItems.splice(1, 0, switchCompany)
    }

    return (
      <header className={styles.headerContainer}>
        <div className={styles.headerNav}>
          <div className={styles.headerLogo} />
          <NavLink to={'/dashboard'} activeClassName={styles.headerActiveTab}>Dashboard</NavLink>
          <NavLink to={'/tasks'} activeClassName={styles.headerActiveTab}>Tasks</NavLink>
        </div>
        <div className={styles.headerInfo}>
          <form className={styles.headerSearchForm}>
            {/* <AutoCompleteInput suggestions={this.taskStore.fullTaskList} icon={<FaSearch/>} placeholder={'Title, checklist, description, id...'}/> */}
          </form>
          <div className={styles.headerProfile}>
            {/* <LocationMenu
              top={100}
              right={true}
              heightAuto={true}
              fullHover={true}
              autoClose={true}
              trigger={this.sliderSelector()}
              items={headerProfileMenuItems}
              callback={this.handleSelection}
              childrenKey={'items'} /> */}
          </div>
        </div>
        {/* <LegalTerms /> */}
        { this.redirect && <Redirect to={this.redirect} /> }
      </header>
    )
  }
}))

export default Header
