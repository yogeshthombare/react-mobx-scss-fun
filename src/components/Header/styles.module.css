@keyframes rotate {
  from {
    transform: rotate(0deg); }
  to {
    transform: rotate(360deg); } }

@keyframes scaleAnim {
  from {
    transform: scale(0.5); }
  to {
    transform: scale(1); } }

.headerContainer {
  background: #207CA7;
  padding: 0 20px 0;
  min-height: 100px;
  display: flex;
  flex-direction: row;
  justify-content: space-between; }
  .headerContainer .headerNav {
    display: flex;
    align-items: center;
    transition: all 0.3s ease-in-out; }
    .headerContainer .headerNav .headerLogo {
      background: url("../../assets/images/logo.svg") no-repeat center/contain;
      width: 220px;
      height: 60px;
      margin-right: 30px; }
    .headerContainer .headerNav a, .headerContainer .headerNav :not(.headerLogo) {
      color: white;
      text-decoration: none;
      font-size: 18px;
      margin: 0 30px;
      transition: all 0.3s ease-in-out;
      align-items: center;
      display: flex;
      border-bottom: 4px solid #207CA7;
      height: auto;
      padding: 4px 0; }
      .headerContainer .headerNav a:first-child, .headerContainer .headerNav :not(.headerLogo):first-child {
        margin: 0 30px 0 0; }
      .headerContainer .headerNav a:not(:first-child):hover, .headerContainer .headerNav :not(.headerLogo):not(:first-child):hover {
        opacity: 0.7; }
    .headerContainer .headerNav .headerActiveTab {
      border-bottom: 4px solid white; }
  .headerContainer .headerInfo {
    display: flex;
    align-items: center;
    justify-content: flex-end; }
    .headerContainer .headerInfo .sliderMenuLogout {
      cursor: pointer;
      color: #db2828; }
      .headerContainer .headerInfo .sliderMenuLogout:hover {
        transition: all 0.3s ease-in-out;
        background: #db2828 !important;
        color: white; }
    .headerContainer .headerInfo .sliderMenuVersion {
      position: absolute;
      display: flex;
      top: 0;
      left: 0;
      right: 0;
      bottom: -1px;
      align-items: center;
      padding: 0 20px;
      cursor: default;
      pointer-events: none;
      color: #bababc;
      background-color: white; }
  .headerContainer .headerSearchForm {
    display: flex;
    align-items: center;
    justify-content: center;
    min-width: 250px;
    margin-right: 50px; }
    .headerContainer .headerSearchForm label {
      margin: 0;
      padding: 0; }
  .headerContainer .headerProfile {
    cursor: pointer;
    display: flex;
    height: 100%;
    flex-direction: row;
    align-items: center;
    position: relative; }
  .headerContainer .headerProfileImage {
    min-width: 60px;
    height: 60px;
    border-radius: 60px;
    background: white;
    margin-right: 20px;
    display: flex;
    align-items: center;
    justify-content: center; }
    .headerContainer .headerProfileImage h2 {
      color: #363f50;
      font-weight: 600;
      text-transform: uppercase;
      letter-spacing: -2px;
      font-size: 24px; }
  .headerContainer .headerProfileInfo {
    display: flex;
    flex-direction: column;
    justify-content: center; }
    .headerContainer .headerProfileInfo h5, .headerContainer .headerProfileInfo p {
      margin: 0;
      padding: 0;
      color: white; }
    .headerContainer .headerProfileInfo p {
      font-weight: lighter;
      font-size: 12px; }
