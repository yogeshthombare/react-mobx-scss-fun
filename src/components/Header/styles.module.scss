@import '../../common/variables.scss';

@mixin takeOver() {
  position: absolute;
  display: flex;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1px;
  align-items: center;
  padding: 0 20px;
}

.headerContainer {
  background: $blue;
  padding: 0 20px 0;
  min-height: 100px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;

  .headerNav {
    display: flex;
    align-items: center;
    transition: all 0.3s ease-in-out;

    .headerLogo {
      background: url('../../assets/images/logo.svg') no-repeat center/contain;
      width: 220px;
      height: 60px;
      margin-right: 30px;
    }

    a, :not(.headerLogo) {
      color: white;
      text-decoration: none;
      font-size: 18px;
      margin: 0 30px;
      transition: all 0.3s ease-in-out;
      align-items: center;
      display: flex;
      border-bottom: 4px solid $blue;
      height: auto;
      padding: 4px 0;

      &:first-child {
        margin: 0 30px 0 0;
      }

      &:not(:first-child):hover {
        opacity: 0.7;
      }
    }

    .headerActiveTab {
      border-bottom: 4px solid white;
    }
  }

  .headerInfo {
    display: flex;
    align-items: center;
    justify-content: flex-end;

    .sliderMenuLogout {
      cursor: pointer;
      color: $error-red;

      &:hover {
        transition: all 0.3s ease-in-out;
        background: $error-red !important;
        color: white;
      }
    }

    .sliderMenuVersion {
      @include takeOver;
      cursor: default;
      pointer-events: none;
      color: $gray-darker;
      background-color: white;
    }
  }

  .headerSearchForm {
    display: flex;
    align-items: center;
    justify-content: center;
    min-width: 250px;
    margin-right: 50px;

    label {
      margin: 0;
      padding: 0;
    }
  }

  .headerProfile {
    cursor: pointer;
    display: flex;
    height: 100%;
    flex-direction: row;
    align-items: center;
    position: relative;
  }

  .headerProfileImage {
    min-width: 60px;
    height: 60px;
    border-radius: 60px;
    background: white;
    margin-right: 20px;
    display: flex;
    align-items: center;
    justify-content: center;

    h2 {
      color: $blue-dark;
      font-weight: 600;
      text-transform: uppercase;
      letter-spacing: -2px;
      font-size: 24px;
    }
  }

  .headerProfileInfo {
    display: flex;
    flex-direction: column;
    justify-content: center;

    h5, p {
      margin: 0;
      padding: 0;
      color: white
    }

    p {
      font-weight: lighter;
      font-size: 12px;
    }
  }
}
