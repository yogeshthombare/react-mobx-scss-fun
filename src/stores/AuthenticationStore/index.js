import { types, flow, getParent } from 'mobx-state-tree'
import * as cookies from 'browser-cookies'
// import { User } from '../../models/UserModel'

export const AuthenticationStore = types
  .model('AuthenticationStore', {
    user: types.frozen(), // FIXME: types.maybe(User) this needs to be updated once the response shape can be defined as JSON
    loading: types.optional(types.boolean, false),
    error: types.frozen(),
    message: types.frozen(),
    authenticated: types.optional(types.boolean, false),
    privacyStatement: types.optional(types.string, ''),
  })
  .actions(self => ({
    fetchCurrentUser: flow(function * () {
      // let currentToken = fetchToken()
      // self.error = undefined
      // self.message = undefined
      // self.loading = true
      // self.authenticated = !!currentToken.token
    }),
  //     if (!currentToken.token) {
  //       self.user = undefined
  //       self.loading = false
  //       return
  //     }

  //     try {
  //       const res = yield fetch(`${self.utility.baseUrl}/api/current_user?permissions_format=nested`, self.utility.setRequest())
  //       const response = yield res.json()
  //       if (response.error) {
  //         self.utility.handleError(response.error)
  //         self.authenticated = false
  //         return
  //       }

  //       self.user = response.data
  //       self.authenticated = true
  //       self.loading = false

  //       self.fetchPrivacyStatement()
  //       self.utility.fetchReleaseNotes()
  //       self.teamStore.fetchTeams()
  //       self.mapStore.fetchPinTypes()
  //     } catch(error) {
  //       let notification = {
  //         message: error.error,
  //         error: true
  //       }
  //       self.utility.addNotification(notification)
  //       self.error = error
  //       self.loading = false
  //     }
  //   }),

  //   fetchPrivacyStatement: flow(function * () {
  //     self.error = undefined
  //     self.message = undefined
  //     self.loading = true

  //     const team_id = self.user.team.id
  //     const url = `${self.utility.baseUrl}/api/teams/${team_id}/privacy_statements/app?format=html`
  //     try {
  //       const res = yield fetch(url, self.utility.setRequest())
  //       self.privacyStatement = yield res.text()
  //       self.loading = false
  //     } catch(error) {
  //       self.handleError(error)
  //     }
  //   }),

  //   submitLogin: flow(function * (email, password) {
  //     self.error = undefined
  //     self.loading = true
  //     try {
  //       const res = yield fetch(`${self.utility.baseUrl}/api/sign_in`, {
  //         method: 'post',
  //         mode: 'cors',
  //         headers: { 'Content-Type': 'application/json' },
  //         body: JSON.stringify({ user: {email, password}})
  //       })

  //       const response = yield res.json()

  //       if (response.error) {
  //         let notification = {
  //           message: response.error,
  //           error: true
  //         }
  //         self.loading = false
  //         self.utility.addNotification(notification)
  //         self.error = response.error
  //         return
  //       } else {
  //         generateAccessToken(email, response.data.authentication_token)
  //         self.fetchCurrentUser()
  //       }

  //       self.user = response.data
  //       self.authenticated = true
  //       self.loading = false
  //     } catch(error) {
  //       let notification = {
  //         message: error.error,
  //         error: true
  //       }
  //       self.utility.addNotification(notification)
  //       self.error = error
  //       self.loading = false
  //     }
  //   }),

  //   submitLogout: flow(function * () {
  //     let currentToken = fetchToken()
  //     self.error = undefined
  //     self.user = undefined
  //     self.authenticated = false
  //     self.loading = true
  //     try {
  //       yield fetch(`${self.utility.baseUrl}/api/sign_out`, {
  //         method: 'delete',
  //         mode: 'cors',
  //         headers: {
  //           'Content-Type': 'application/json',
  //           'X-USER-EMAIL': currentToken.email,
  //           'X-USER-TOKEN': currentToken.token
  //         }
  //       })

  //       cookies.erase('token')
  //       cookies.erase('email')
  //       // self.taskStore.removeGlobalFilter(null, true)
  //       self.mapStore.updateCurrentMap(null, true)
  //       self.app.resetApplication()
  //       self.loading = false
  //     } catch(error) {
  //       self.handleError(error)
  //     }
  //   }),

  //   submitAccountRecovery: flow(function * (email) {
  //     self.error = undefined
  //     self.loading = true
  //     try {
  //       const res = yield fetch(`${self.utility.baseUrl}/api/forgot_password`, {
  //         method: 'post',
  //         mode: 'cors',
  //         headers: {
  //           'Content-Type': 'application/json',
  //         },
  //         body: JSON.stringify({
  //           'email': email
  //         })
  //       })

  //       const response = yield res.json()

  //       if (response.error || response.message) {
  //         self.handleError(response.error || response.message)
  //         return
  //       }

  //       self.authenticated = false
  //       self.loading = false
  //     } catch(error) {
  //       self.handleError(error)
  //     }
  //   }),

  //   // FIXME: when everyone decides what they want this to do
  //   submitAccountCreation(params) {
  //     console.log('submitAccountCreation:', params)
  //   },

  //   submitPasswordReset(params) {
  //     console.log('submitPasswordReset:', params)
  //   },
  //   //end FIXME

    handleError(error) {
      console.log(error)
      let notification = { message: error.error, error: true }
      self.utility.addNotification(notification)
      self.error = error
      self.loading = false
    },
  }))
  .views(self => ({
    get userTeams() {
      return self.user.teams.map(item => {
        return item.name
      })
    },

    get utility() {
      return getParent(self).utilityStore
    },

    get mapStore() {
      return getParent(self).mapStore
    },

    get taskStore() {
      return getParent(self).taskStore
    },

    get projectStore() {
      return getParent(self).projectStore
    },

    get teamStore() {
      return getParent(self).teamStore
    },

    get app() {
      return getParent(self)
    }
  }))

const generateAccessToken = (email, token) => {
  cookies.set('token', token)
  cookies.set('email', email)
}

const fetchToken = () => {
  let token = cookies.get('token')
  let email = cookies.get('email')
  return({
    'token': token,
    'email': email
  })
}
