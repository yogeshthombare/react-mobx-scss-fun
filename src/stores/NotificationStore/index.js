import { types } from 'mobx-state-tree'
export const NotificationStore = types
  .model('NotificationStore', {
    showFeed: types.frozen(true),
    showCards: types.frozen(true),
    currentPage: types.optional(types.string, 'dashboard')
  })
  .actions(self => ({
    toggleDashboardCards() {
      self.showCards = !self.showCards
    },
    toggleActivity() {
      self.showFeed = !self.showFeed
    },
  }))
  .views(self => ({

  }))
