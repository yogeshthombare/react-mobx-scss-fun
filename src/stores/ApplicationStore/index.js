import { types, getSnapshot, applySnapshot } from 'mobx-state-tree'
import { AuthenticationStore } from '../AuthenticationStore'


let initialState = {}

export const ApplicationStore = types
  .model('ApplicationStore', {
    auth: types.optional(AuthenticationStore, {
      user: undefined,
      loading: false
    }),
    // taskStore: types.optional(TaskStore, {
    //   tasks: [],
    //   filteredTasks: []
    // }),
    // projectStore: types.optional(ProjectStore, {
    //   projects: []
    // }),
    // utilityStore: types.optional(UtilityStore, {
    //   notifications: []
    // }),
    // mapStore: types.optional(MapStore, {
    //   activeMap: undefined
    // }),
    // filterStore: types.optional(FilterStore, {
    //   pins: [],
    //   teamUsers: [],
    //   teamPriorities: [],
    //   filterMenuItems: [],
    //   dueDate: undefined
    // }),
    // notificationStore: types.optional(NotificationStore, {
    //   notificationsFeed: [],
    //   showFeed: false,
    //   allNotifications: false,
    // }),
    // activityFeedStore: types.optional(ActivityFeedStore, {
    //   showFeed: true,
    // }),
    // teamStore: types.optional(TeamStore, {
    //   teams: [],
    //   activeTeam: {}
    // }),
    // reportStore: types.optional(ReportStore, {
    //     dashboardCards: [],
    //     showStats: true,
    // }),
    // dashboardStore: types.optional(DashboardStore, {
    //     tasks: [],
    //     filteredTasks: [],
    // }),
    // taskboardStore: types.optional(TaskboardStore, {
    //   tasks: [],
    //   filteredTasks: []
    // }),
    currentPage: types.optional(types.string, 'dashboard')
  })
  .actions(self => ({

    afterCreate() {
      initialState = getSnapshot(self)
    },

    resetApplication() {
      applySnapshot(self, initialState)
      self.updateCurrentPage('dashboard')
    },

    updateCurrentPage(page) {
      self.currentPage = page
    }
  }))
  .views(self => ({
    get currentlyActivePage() {
      return self.currentPage
    }
  }))
