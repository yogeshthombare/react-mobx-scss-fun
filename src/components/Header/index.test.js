import React from 'react'
import ReactDOM from 'react-dom'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Header from '../Header'
import { ApplicationStore } from '../../stores/ApplicationStore'

let app = ApplicationStore.create()

configure({
  adapter: new Adapter()
})

describe('Header', () => {
  it('component is loaded and renders as expected', () => {
    const component = shallow(<Header app={app}/>)
    expect(component).toMatchSnapshot()
  })
})
