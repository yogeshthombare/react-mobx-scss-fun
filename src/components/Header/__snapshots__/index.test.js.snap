// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component is loaded and renders as expected 1`] = `ShallowWrapper {}`;
